### ISSUE TITLE
unable to get remote tags

### ISSUE THREAD

**Issue by thinkrapido** [nim-lang/nimble#204]
Hi,

i successfully installed nimble.

then I tried to install nimx.

Everything goes well 'til this point:
Error: unhandled exception: Unable to query remote tags for https://github.com/fowlmouth/nake/. Git returned: ERROR: Repository not found.

How can I fix this?

**Comment by dom96** [comment]
What version of git do you have?

**Comment by thinkrapido** [comment]
2.1.4

**Comment by dom96** [comment]
Please run ``git ls-remote --tags https://github.com/fowlmouth/nake/`` and show me the output.

Also, what OS are you on?

**Comment by thinkrapido** [comment]
> ssh: connect to host github.com port 22: Connection refused
> fatal: Could not read from remote repository.
> 
> Please make sure you have the correct access rights
> and the repository exists.

os is ubuntu 15.04

**Comment by thinkrapido** [comment]
i cannot clone git repositories by https instead i always use ssh

**Comment by dom96** [comment]
How come?

**Comment by thinkrapido** [comment]
I don't know.
Maybe I have a misconfigured .gitconfig file

**Comment by thinkrapido** [comment]
After several tries and nothing changed the nimble install nimx worked and I could launch my first up.

I really don't know what went wrong!

Thanks anyway

**Comment by dom96** [comment]
Bad internet maybe? I will close this in any case, reopen if it keeps happening.

**Comment by Jeff-Ciesielski** [comment]
I'm experiencing the same issue when attempting to install nimx/nake on osx.

Output from git ls-remote:

~ ❯❯❯ git ls-remote --tags https://github.com/fowlmouth/nake                                                                                                                       ⏎
ebe39f691f069b374c7b2e3ab12db52d7b15e697	refs/tags/v1.0
6040b3f30cd89950402c1bf538ed53d8a48a6524	refs/tags/v1.1
509eb8200c84826a74314cffd67c59b968350395	refs/tags/v1.2
46ce7196852f22e106c066f5370c6b0c4bf455ea	refs/tags/v1.2^{}
0b23e746658e0711584f5c3ba782cba614ebd188	refs/tags/v1.4
f8a78e3c9fdf88595867b7d44b3fc653cbd3a0e2	refs/tags/v1.4^{}
112f48c2b6a3a3af429938299ed3923e1d23f59f	refs/tags/v1.6
3fe61c7dc8db7881ebbcf9ceb7f77e4fd6c8dc22	refs/tags/v1.6^{}
b6e82f4aeaca88494cfe34fd4f2e726cc33fd86b	refs/tags/v1.8
a863d7929344ffdf1d6354e709641cc2abfeb3d0	refs/tags/v1.8^{}


Output from Nimble:

~ ❯❯❯ nimble install nake                                                                                                                                                          ⏎
Searching in "official" package list...
Downloading https://github.com/fowlmouth/nake/ into /tmp/nimble_8703/githubcom_fowlmouthnake using git...
Traceback (most recent call last)
nimble.nim(962)          nimble
nimble.nim(894)          doAction
nimble.nim(536)          install
nimble.nim(495)          downloadPkg
download.nim(191)        doDownload
download.nim(91)         getTagsListRemote
Error: unhandled exception: Unable to query remote tags for https://github.com/fowlmouth/nake/. Git returned: ERROR: Repository not found.
fatal: Could not read from remote repository.


As per OP's fix, tried reinstalling several times, no bueno.

**Comment by dom96** [comment]
What version of git do you have?

**Comment by Jeff-Ciesielski** [comment]
2.5.4 (Apple Git-61)

**Comment by dom96** [comment]
Can you try ``git ls-remote --tags https://github.com/fowlmouth/nake/`` (note the ``/`` at the end)? @boydgreenfield reported this as an error, seems like that might be it.

**Comment by Jeff-Ciesielski** [comment]
Yep, that gives a repo not found error!

**Comment by dom96** [comment]
Very weird. I just tried the same command on my MacBook with the same version of Git and it works both with the `/` and without.

**Comment by Jeff-Ciesielski** [comment]
Just reproduced with git 2.7.2 on OSX.

Additional Info: Only happens with HTTPS, amending the URL to `http://github.com/fowlmouth/nake/` pulls the tags just fine.  As a workaround, I was able to change my ~/.gitconfig to rewrite git@github.com urls to http and was able to install nimx/nake, though this seems unacceptable long term.

Double checked against my bitbucket account and this issue isn't present, trailing slashes, http/https all work, so I think this is actually a github issue rather than a problem with git itself.  I've sent an email to their support team and will update this bug when/if I get a response.

**Comment by dom96** [comment]
@Jeff-Ciesielski Thank you for looking into it! This is a very curious bug, excited to hear what Github support will have to say about it :)

**Comment by Jeff-Ciesielski** [comment]
Ok, so I'm towards the end of my rope on this one.  I *think* it's github's issue, but I don't expect that they'll be much more help.  

Their tech support got back to me and asked me to run with CURL_VERBOSE turned on.  Doing so:

- works fine on all bitbucket repos, and HTTP only github repos.  
- Github HTTPS with a trailing slash fails outright
- Github HTTPS without a trailing slash will hit the repository, but no additional verbose info is printed out (weird).

When I informed them that I was on osx 10.10.5 (Yosemite) and was thinking about upgrading to see if it improved the situation, I was met with a response of "Oh, that's why we can't repro, we're on ElCapitan! Let us know how upgrading goes"... :\. 

So TL;DR: I think the safe bet (which you also alluded to in [this bug](https://github.com/nim-lang/packages/pull/319)) to just perform a trailing slash strip in the tool itself.  That said, AFAIK, a url with a trailing slash is actually distinct from one without, so this isn't *technically* correct.  Maybe best to only do this on github URLs to avoid (potentially) breaking other git hosting services (in case they make use of this behavior)?

I'd be happy to take a crack at it and submit a PR if you're busy (Fairly new to Nim, but I dont think a change like that is rocket surgery!)

**Comment by dom96** [comment]
There must be something that changed between Yosemite and El Capitan. Because I am running El Capitan and it works for me. I appreciate you looking into it. Have you been able to figure out how Git accesses those URLs, is it using curl? Maybe you could try other curl-like software and see what results it gives.

I would be more than happy to accept a PR from you with the fix you described.

**Comment by dom96** [comment]
Just for further reference, this commit works around this bug https://github.com/nim-lang/nimble/commit/e51d1679c5549a5775bc8a6e185b2a7116c15978

**Comment by dom96** [comment]
Going to close this as I don't think it's an issue any longer.
